# Lab 3 - Multiparticle Systems

In [1]:
from pylab import *

Functions to define spin-$s$ matrices:

In [2]:
def spinx(s):

    n = int(2.0*s+1)
    sx = matrix(zeros((n,n)))
    for a in range(0,n):
        for b in range(0,n):
            if (a==b+1):
                sx[a,b] = sx[a,b] + 0.5*sqrt((s+1)*(a+b+1)-(a+1)*(b+1))
            elif (a==b-1):
                sx[a,b] = sx[a,b] + 0.5*sqrt((s+1)*(a+b+1)-(a+1)*(b+1))
                
    return sx


def spiny(s):
    n = int(2.0*s+1)
    sy = matrix(zeros((n,n),dtype='complex'))

    for a in range(0,n):
        for b in range(0,n):
            if (a==b+1):
                sy[a,b] = sy[a,b] + 0.5j*sqrt((s+1)*(a+b+1)-(a+1)*(b+1))
            elif (a==b-1):
                sy[a,b] = sy[a,b] - 0.5j*sqrt((s+1)*(a+b+1)-(a+1)*(b+1))
                
    return sy
    
def spinz(s):
    
    n = int(2.0*s+1)
    sz = matrix(zeros((n,n)))

    for a in range(0,n):
        for b in range(0,n):
    
            if (a==b):
                sz[a,b] = (s+1-b-1)
                
    return sz

# 1.

In [3]:
s = 2
print(spinx(s))
print(spiny(s))
print(spinz(s))

[[0.         1.         0.         0.         0.        ]
 [1.         0.         1.22474487 0.         0.        ]
 [0.         1.22474487 0.         1.22474487 0.        ]
 [0.         0.         1.22474487 0.         1.        ]
 [0.         0.         0.         1.         0.        ]]
[[0.+0.j         0.-1.j         0.+0.j         0.+0.j
  0.+0.j        ]
 [0.+1.j         0.+0.j         0.-1.22474487j 0.+0.j
  0.+0.j        ]
 [0.+0.j         0.+1.22474487j 0.+0.j         0.-1.22474487j
  0.+0.j        ]
 [0.+0.j         0.+0.j         0.+1.22474487j 0.+0.j
  0.-1.j        ]
 [0.+0.j         0.+0.j         0.+0.j         0.+1.j
  0.+0.j        ]]
[[ 2.  0.  0.  0.  0.]
 [ 0.  1.  0.  0.  0.]
 [ 0.  0.  0.  0.  0.]
 [ 0.  0.  0. -1.  0.]
 [ 0.  0.  0.  0. -2.]]


In [4]:
x = spinx(s)*spiny(s)-spiny(s)*spinx(s)
print(x)

[[0.+2.00000000e+00j 0.+0.00000000e+00j 0.+0.00000000e+00j
  0.+0.00000000e+00j 0.+0.00000000e+00j]
 [0.+0.00000000e+00j 0.+1.00000000e+00j 0.+0.00000000e+00j
  0.+0.00000000e+00j 0.+0.00000000e+00j]
 [0.+0.00000000e+00j 0.+0.00000000e+00j 0.-8.71111819e-17j
  0.+0.00000000e+00j 0.+0.00000000e+00j]
 [0.+0.00000000e+00j 0.+0.00000000e+00j 0.+0.00000000e+00j
  0.-1.00000000e+00j 0.+0.00000000e+00j]
 [0.+0.00000000e+00j 0.+0.00000000e+00j 0.+0.00000000e+00j
  0.+0.00000000e+00j 0.-2.00000000e+00j]]


The commutator comes out to be 2i,1i,0,-1i,-2i along the identity! For SxSy, we get iSz! Cool

In [5]:
x = spiny(s)*spinz(s)-spinz(s)*spiny(s)
print(x)

[[0.+0.j         0.+1.j         0.+0.j         0.+0.j
  0.+0.j        ]
 [0.+1.j         0.+0.j         0.+1.22474487j 0.+0.j
  0.+0.j        ]
 [0.+0.j         0.+1.22474487j 0.+0.j         0.+1.22474487j
  0.+0.j        ]
 [0.+0.j         0.+0.j         0.+1.22474487j 0.+0.j
  0.+1.j        ]
 [0.+0.j         0.+0.j         0.+0.j         0.+1.j
  0.+0.j        ]]


SySz comes out to be iSx!

In [6]:
S2 = spinx(s)**2 + spiny(s)**2 + spinz(s)**2
S2

matrix([[6.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j],
        [0.+0.j, 6.+0.j, 0.+0.j, 0.+0.j, 0.+0.j],
        [0.+0.j, 0.+0.j, 6.+0.j, 0.+0.j, 0.+0.j],
        [0.+0.j, 0.+0.j, 0.+0.j, 6.+0.j, 0.+0.j],
        [0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 6.+0.j]])

General result is entirely real 6 along the identity!

In [7]:
s2 = 3.0/2.0
print(spinx(s2))
print(spiny(s2))
print(spinz(s2))
S2 = spinx(s2)**2 + spiny(s2)**2 + spinz(s2)**2
print(S2)

[[0.        0.8660254 0.        0.       ]
 [0.8660254 0.        1.        0.       ]
 [0.        1.        0.        0.8660254]
 [0.        0.        0.8660254 0.       ]]
[[0.+0.j        0.-0.8660254j 0.+0.j        0.+0.j       ]
 [0.+0.8660254j 0.+0.j        0.-1.j        0.+0.j       ]
 [0.+0.j        0.+1.j        0.+0.j        0.-0.8660254j]
 [0.+0.j        0.+0.j        0.+0.8660254j 0.+0.j       ]]
[[ 1.5  0.   0.   0. ]
 [ 0.   0.5  0.   0. ]
 [ 0.   0.  -0.5  0. ]
 [ 0.   0.   0.  -1.5]]
[[3.75+0.j 0.  +0.j 0.  +0.j 0.  +0.j]
 [0.  +0.j 3.75+0.j 0.  +0.j 0.  +0.j]
 [0.  +0.j 0.  +0.j 3.75+0.j 0.  +0.j]
 [0.  +0.j 0.  +0.j 0.  +0.j 3.75+0.j]]


The S2 here is 3.75 along the identity. This is not the same as the S2 operator for s=2. But they are both on the identity! This proportionality follows S^2  + S

In [8]:
s3 = 1/2
print(spinx(s3)**2)
print(spiny(s3)**2)
print(spinz(s3)**2)

[[0.25 0.  ]
 [0.   0.25]]
[[0.25+0.j 0.  +0.j]
 [0.  +0.j 0.25+0.j]]
[[0.25 0.  ]
 [0.   0.25]]


They are all 1/4 I for s=1/2, but for other s...

In [9]:
s4 = 1
print(spinx(s4)**2)
print(spiny(s4)**2)
print(spinz(s4)**2)

[[0.5 0.  0.5]
 [0.  1.  0. ]
 [0.5 0.  0.5]]
[[ 0.5+0.j  0. +0.j -0.5+0.j]
 [ 0. +0.j  1. +0.j  0. +0.j]
 [-0.5+0.j  0. +0.j  0.5+0.j]]
[[1. 0. 0.]
 [0. 0. 0.]
 [0. 0. 1.]]


They are not the same! The pattern is very different.

# --------------------------------------------------

In [10]:
u1 = array([1.0,0.0])
u1.shape = (2,1)
u2 = array([0.0,1.0])
u2.shape = (2,1)

print(np.kron(u1,u1))
print(np.kron(u1,u2))
print(np.kron(u2,u1))
print(np.kron(u2,u2))
p1 = np.kron(u1,u1)
p2 = np.kron(u1,u2)
p3 = np.kron(u2,u1)
p4 = np.kron(u2,u2)

[[1.]
 [0.]
 [0.]
 [0.]]
[[0.]
 [1.]
 [0.]
 [0.]]
[[0.]
 [0.]
 [1.]
 [0.]]
[[0.]
 [0.]
 [0.]
 [1.]]


order matters!

In [11]:
so = eye(2)
s5 = 0.5
Sz = kron(spinz(s5),so)+kron(so,spinz(s5))
Sz

matrix([[ 1.,  0.,  0.,  0.],
        [ 0.,  0.,  0.,  0.],
        [ 0.,  0.,  0.,  0.],
        [ 0.,  0.,  0., -1.]])

In [12]:
print(asscalar(dot(conj(p1.transpose()),Sz*p1)))
print(asscalar(dot(conj(p2.transpose()),Sz*p2)))
print(asscalar(dot(conj(p3.transpose()),Sz*p3)))
print(asscalar(dot(conj(p4.transpose()),Sz*p4)))

1.0
0.0
0.0
-1.0


In [13]:
Pi = matrix(zeros((4,4)))
Pi[0,0] = 1
Pi[3,3] = 1
Pi[1,2] = 1
Pi[2,1] = 1
Pi

matrix([[1., 0., 0., 0.],
        [0., 0., 1., 0.],
        [0., 1., 0., 0.],
        [0., 0., 0., 1.]])

In [14]:
Pi**2

matrix([[1., 0., 0., 0.],
        [0., 1., 0., 0.],
        [0., 0., 1., 0.],
        [0., 0., 0., 1.]])

This is the identity matrix!

In [15]:
eigh(Pi)

(array([-1.,  1.,  1.,  1.]),
 matrix([[ 0.        ,  0.        ,  1.        ,  0.        ],
         [-0.70710678,  0.70710678,  0.        ,  0.        ],
         [ 0.70710678,  0.70710678,  0.        ,  0.        ],
         [ 0.        ,  0.        ,  0.        ,  1.        ]]))

Eigenstates of Pi:

-1 (0 |+>|+> - 1/sqrt(2) |+>|-> + 1/sqrt(2) |->|+> + 0 |->|->)

1 (0 |+>|+> + 1/sqrt(2) |+>|-> + 1/sqrt(2) |->|+> + 0 |->|->)

1 ( |+>|+> + 0 |+>|-> + 0 |->|+> + 0 |->|->)

1 (0 |+>|+> + 0 |+>|-> + 0 |->|+> + |->|->)

wow!

In [16]:
s = 0.5
Sx = np.kron(spinx(s),so)+np.kron(so,spinx(s))
Sy = np.kron(spiny(s),so)+np.kron(so,spiny(s))
Sz = np.kron(spinz(s),so)+np.kron(so,spinz(s))

In [17]:
S2 = Sx**2 + Sy**2 + Sz**2
print(S2)
print(eigh(S2))

[[2.+0.j 0.+0.j 0.+0.j 0.+0.j]
 [0.+0.j 1.+0.j 1.+0.j 0.+0.j]
 [0.+0.j 1.+0.j 1.+0.j 0.+0.j]
 [0.+0.j 0.+0.j 0.+0.j 2.+0.j]]
(array([0., 2., 2., 2.]), matrix([[ 0.        +0.j,  0.        +0.j,  1.        +0.j,
          0.        +0.j],
        [-0.70710678+0.j,  0.70710678+0.j,  0.        +0.j,
          0.        +0.j],
        [ 0.70710678+0.j,  0.70710678+0.j,  0.        +0.j,
          0.        +0.j],
        [ 0.        +0.j,  0.        +0.j,  0.        +0.j,
          1.        +0.j]]))


The eigenvalues for S2 for s=1/2 are 0,2,2,2. This means that s is nondefinite, s=-2,1

The eigenvectors of S2 are the same as Pi!!!!

The two singlets differ in that for the first, there is a balance in the probability in finding the states |+>|-> and |->|+>, whereas in the second singlet they both have an equal probability of being observed.

In [18]:
Sz1 = np.kron(spinz(s),so)
Sz2 = np.kron(so,spinz(s))
SzSz = np.kron(spinz(s),spinz(s))
print(Sz1,"\n",Sz2,"\n\n",SzSz)

[[ 0.5  0.   0.   0. ]
 [ 0.   0.5  0.   0. ]
 [ 0.   0.  -0.5 -0. ]
 [ 0.   0.  -0.  -0.5]] 
 [[ 0.5  0.   0.   0. ]
 [ 0.  -0.5  0.  -0. ]
 [ 0.   0.   0.5  0. ]
 [ 0.  -0.   0.  -0.5]] 

 [[ 0.25  0.    0.    0.  ]
 [ 0.   -0.25  0.   -0.  ]
 [ 0.    0.   -0.25 -0.  ]
 [ 0.   -0.   -0.    0.25]]


In [19]:
a = p1
print(a)
b = (1/np.sqrt(2))*(p2-p3)
print(b)
c = (1/np.sqrt(2))*(p2+p3)
print(c)
d = p4
print(p4)

[[1.]
 [0.]
 [0.]
 [0.]]
[[ 0.        ]
 [ 0.70710678]
 [-0.70710678]
 [ 0.        ]]
[[0.        ]
 [0.70710678]
 [0.70710678]
 [0.        ]]
[[0.]
 [0.]
 [0.]
 [1.]]


# a

In [20]:
print(asscalar(dot(conj(a.transpose()),Sz1*a)))
print(asscalar(dot(conj(a.transpose()),Sz2*a)))
print(asscalar(dot(conj(a.transpose()),SzSz*a)))

0.5
0.5
0.25


# b

In [21]:
print(asscalar(dot(conj(b.transpose()),Sz1*b)))
print(asscalar(dot(conj(b.transpose()),Sz2*b)))
print(asscalar(dot(conj(b.transpose()),SzSz*b)))

0.0
0.0
-0.24999999999999994


# c

In [22]:
print(asscalar(dot(conj(c.transpose()),Sz1*c)))
print(asscalar(dot(conj(c.transpose()),Sz2*c)))
print(asscalar(dot(conj(c.transpose()),SzSz*c)))

0.0
0.0
-0.24999999999999994


# D

In [23]:
print(asscalar(dot(conj(d.transpose()),Sz1*d)))
print(asscalar(dot(conj(d.transpose()),Sz2*d)))
print(asscalar(dot(conj(d.transpose()),SzSz*d)))

-0.5
-0.5
0.25


They don't really seem entangled. I don't know how to tell though